test/integration/controller_binding_test.go

```diff
@@ -0,0 +1,554 @@
+				instance: getTestInstance(),
+				binding:  getTestBinding(),
+			}
+			ct.run(func(ct *controllerTest) {
+				{
+					condition := v1beta1.ServiceBindingCondition{
+						Type:   v1beta1.ServiceBindingConditionReady,
+						Status: v1beta1.ConditionTrue,
+					}
+					if cond, err := util.WaitForBindingConditionLastSeenOfType(ct.client, testNamespace, testBindingName, condition); err != nil {
+						t.Fatalf("error waiting for binding condition: %v\n"+"expecting: %+v\n"+"last seen: %+v", err, condition, cond)
+					}
+				}
+			})
+		})
+	}
+}
+
+// TestCreateServiceBindingInvalidInstance try to bind to invalid service instance names
+func TestCreateServiceBindingInvalidInstance(t *testing.T) {
+	cases := []struct {
+		name                string
+		instanceName        *string
+		expectedErrorReason string
+		expectedFailure     bool
+	}{
+		{
+			name:                "non-existent service instance name",
+			instanceName:        strPtr("nothereinstance"),
+			expectedErrorReason: "ReferencesNonexistentInstance",
+		},
+		{
+			name:                "invalid service instance name",
+			instanceName:        strPtr(""),
+			expectedErrorReason: "ReferencesNonexistentInstance",
+			expectedFailure:     true,
+		},
+	}
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			ct := &controllerTest{
+				t: t,
+				skipBindingCreateError: tc.expectedFailure,
+				broker:                 getTestBroker(),
+				instance:               getTestInstance(),
+				binding: func() *v1beta1.ServiceBinding {
+					b := getTestBinding()
+					if tc.instanceName != nil {
+						b.Spec.ServiceInstanceRef.Name = *tc.instanceName
+					}
+					return b
+				}(),
+				skipVerifyingBindingSuccess: tc.expectedErrorReason != "",
+			}
+			ct.run(func(ct *controllerTest) {
+				{
+					status := v1beta1.ConditionTrue
+					if tc.expectedErrorReason != "" {
+						status = v1beta1.ConditionFalse
+					}
+					condition := v1beta1.ServiceBindingCondition{
+						Type:   v1beta1.ServiceBindingConditionReady,
+						Status: status,
+						Reason: tc.expectedErrorReason,
+					}
+					if ct.skipBindingCreateError {
+						if err := util.WaitForBindingToNotExist(ct.client, testNamespace, testBindingName); err != nil {
+							t.Fatalf("error waiting for binding to not exist: %v", err)
+						}
+					} else {
+						if cond, err := util.WaitForBindingConditionLastSeenOfType(ct.client, testNamespace, testBindingName, condition); err != nil {
+							t.Fatalf("error waiting for binding condition: %v\n"+"expecting: %+v\n"+"last seen: %+v", err, condition, cond)
+						}
+					}
+
+				}
+			})
+		})
+	}
+}
+
+// TODO: this one is still broken:
+// E1116 16:10:21.295296    8302 controller_instance.go:1844] ServiceInstance "test-namespace/test-instance": Failed to update status: ServiceInstance.servicecatalog.k8s.io "test-instance" is invalid: status.currentOperation: Forbidden: currentOperation must not be present when reconciledGeneration and generation are equal
```

Comment:
This is a bug in the Controller around deleting a ServiceInstance while there is an async operation in progress. I will write up an issue about it.

Reply:
For reference, #1587.